tools/tests/form-control/spec/test.js

```diff
@@ -0,0 +1,196 @@
+
+import { LOGIN_TEMPLATE, REGISTER_TEMPLATE, PREVENT_SUBMIT_TEMPLATE, CHECKBOXES_TEMPLATE, RADIO_TEMPLATE, SWITCH_TEMPLATE, DROPDOWN_TEMPLATE, } from '../utils/templates';
+import { formsIds, FORM_ADDITIONAL_STYLES, RESPONSE_CONTAINER_ID } from '../utils/constants';
+import { submitForm, checkResponseData } from '../utils/actions';
+
+export const formsTestNames = {
+	LOGIN_FORM: 'Login form',
+	REGISTER_FORM: 'Register form',
+	PREVENT_SUBMIT_FORM: 'Prevent form from submiting',
+	CHECKBOXES_FORM: 'Checkbox cases form',
+	RADIO_FORM: 'Radio cases form',
+	SWITCH_FORM: 'Switch cases form',
+	DROPDOWN_FORM: 'Dropdown cases form',
+}
+
+const forms = [
+	{
+		testName: formsTestNames.LOGIN_FORM,
+		id: formsIds.LOGIN_FORM,
+		submitData: '{"username":"name","password":"pass","poll":"cats"}',
+		submitDataInteraction: '{"username":"name","password":"pass","poll":"dogs"}',
+		template: LOGIN_TEMPLATE,
+	},
+	{
+		testName: formsTestNames.REGISTER_FORM,
+		id: formsIds.REGISTER_FORM,
+		submitData: '{"username":"name","password":"password","info":"user info","gender":"male","user-interests":["music","coding"],"skill-level":"Beginner"}',
+		submitDataInteraction: '{"username":"name","password":"password","info":"user info","gender":"female","user-interests":"coding","skill-level":"Expert"}',
+		template: REGISTER_TEMPLATE,
+	},
+	{
+		testName: formsTestNames.PREVENT_SUBMIT_FORM,
+		id: formsIds.PREVENT_SUBMIT_FORM,
+		shouldPreventSubmit: true,
+		submitData: 'Prevented default',
+		template: PREVENT_SUBMIT_TEMPLATE,
+	},
+	{
+		testName: formsTestNames.CHECKBOXES_FORM,
+		id: formsIds.CHECKBOXES_FORM,
+		submitData: '{"user-interests":["music","on"]}',
+		submitDataInteraction: '{}',
+		template: CHECKBOXES_TEMPLATE,
+	},
+	{
+		testName: formsTestNames.RADIO_FORM,
+		id: formsIds.RADIO_FORM,
+		submitData: '{"option1":"1","option3":"on","option5":"3"}',
+		submitDataInteraction: '{"option1":"2","option3":"on","option5":"3"}',
+		template: RADIO_TEMPLATE,
+	},
+	{
+		testName: formsTestNames.SWITCH_FORM,
+		id: formsIds.SWITCH_FORM,
+		submitData: '{}',
+		submitDataInteraction: '{"option1":"option1-checked","option3":"on"}',
+		template: SWITCH_TEMPLATE,
+	},
+	{
+		testName: formsTestNames.DROPDOWN_FORM,
+		id: formsIds.DROPDOWN_FORM,
+		submitData: '{"option1":"1","option2":"Two","option4":"Two","option5":["1","Two","3"]}',
+		// Expected options will change when the caching feature is reworked or removed.
+		submitDataInteraction: '{"option1":"2","option2":"Two","option4":"Two","option5":["1","3","Two","3"]}',
+		template: DROPDOWN_TEMPLATE,
+	},
+];
+
+async function setupFormControlPage(form) {
+	const el = document.createElement('div');
+	el.className = 'test-wrapper';
+
+	el.innerHTML = form.template;
+
+	cleanTestPage('.test-wrapper');
+
+	document.body.appendChild(el);
+
+	document.getElementById(form.id).addEventListener('loadend', (event) => {
```

Comment:
Aren't we supposed to wait 2 frames before we can access an element that we've just attached to the dom?(in Gameface)

Reply:
Well if we add an event listener that is not required. When `gameface-form-control` it is uninitialized it will be an HTML element and we will add event listener to and HTMLElement. Then it will be initialized with the FormControl class but the eventlister will be still there. I tested this :D

However, I will add it after 2 frames .